### PR TITLE
Document Entity AI

### LINKED ISSUE #377: Explain AI tasks and goals
The AI API is getting finished soon, so we need to document:
- [ ] how it works
- [ ] what tasks are
  - [ ] how you implement your own tasks
- [ ] what goals are
- [ ] how to use/alter the Vanilla AI or a Mod AI
- [ ] how to implement your own AI
- [ ] explain all available events
  - [ ] .AITaskEvent.Add/Remove

### PR DESCRIPTION
**Pull request by ST-DDT**
Fixes #377.

* [x] Explain Agents
* [x] Explain Goals
* [x] Explain Tasks
  * [x] Example change Goal using existing Tasks
  * [x] Example implement your own task
* [x] Explain Events

### CODE REVIEW

**Comment by Spongy** [comment]
A preview for this pull request is available at https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/e530314/.

Here are some links to the pages that were modified:

- added: https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/e530314/plugin/entities/ai.html
- added: https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/e530314/plugin/entities/custom-ai.html
- modified: https://cdn.rawgit.com/Spongy/SpongeDocs-PRs/e530314/plugin/entities/index.html

_Since the preview frequently changes, please link to [this comment](https://github.com/SpongePowered/SpongeDocs/pull/741#issuecomment-412946095), not to the direct url to the preview._

**Comment by ST-DDT** [comment]
Ready for review.

File: source/plugin/entities/ai.rst [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+The Sponge API tries to provides a disambiguator. Calling :javadoc:`AITask#getType()` returns an :javadoc:`AITaskType`
+which can be used to differentiate between the existing types of tasks.
+
+First we try the simple version which will only work if there are no AI altering mods present:
+
+.. code-block:: java
+
+    Goal<Zombie> goal = ...;
+    
+    AITask<Zombie> attackTask = (AITask<Zombie>) goal.getTasks().get(1); // EntityAIZombieAttack
```
Comment (liach) [inline source/plugin/entities/ai.rst:143]: Should be `AITask<? super Zombie>`

File: source/plugin/entities/custom-ai.rst [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+    }
+
+    @Override
+    public void update() {
+    }
+
+    @Override
+    public void reset() {
+        getNavigator().clearPath();
+        this.optTarget = null;
```
Comment (liach) [inline source/plugin/entities/custom-ai.rst]: `this.optTarget = Optional.empty();`

File: source/plugin/entities/custom-ai.rst [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+        if (owner.getRandom().nextFloat() > EXECUTION_CHANCE) {
+            return false;
+        }
+
+        final Vector3d position = getPositionOf(owner);
+        this.optTarget = owner.getWorld()
+                .getEntities().stream()
+                .filter(this.entityFilter)
+                .filter(e -> getPositionOf(e).distanceSquared(position) < MAX_DISTANCE_SQUARED)
+                .sorted(Comparator.comparingDouble(e -> getPositionOf(e).distanceSquared(position)))
```
Comment (liach) [inline source/plugin/entities/custom-ai.rst]: Use `.min(Comparator.comparingDouble(e -> getPositionOf(e).distanceSquared(position)))` instead

File: source/plugin/entities/ai.rst (hunk 1 above)
Comment (ST-DDT) [reply source/plugin/entities/ai.rst:143]: What would be the benefit? I still cannot avoid the cast!
The method returns `? super AITask<? extends Zombie>` aka. `Object`, which is a nuisance.

Comment (liach) [reply source/plugin/entities/ai.rst:143]: No, that method itself is broke (I pointed out before to zidane). The correct type returned should be `<T extends AITask<? super Zombie>>`. For example, if there is a `WalkToTargetAITask extends AITask<Agent>`, `WalkToTargetAITask` should be applicable to a `Zombie`.

**Review by liach — APPROVED** [review]
Anyways, that other generic thing is subject to change in API 8. I believe the generic type needs to be refined.

File: source/plugin/entities/ai.rst (hunk 1 above)
Comment (liach) [reply source/plugin/entities/ai.rst:143]: This is trivial and can be disregarded for now.

File: source/plugin/entities/ai.rst [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+    org.spongepowered.api.entity.ai.task.AITaskType
+    org.spongepowered.api.entity.ai.task.AITaskTypes
+    org.spongepowered.api.event.entity.ai.AITaskEvent.Add
+    org.spongepowered.api.event.entity.ai.AITaskEvent.Remove
+    org.spongepowered.api.event.entity.ai.SetAITargetEvent
+    org.spongepowered.api.entity.ai.task.builtin.WatchClosestAITask
+    org.spongepowered.api.entity.living.Agent
+    org.spongepowered.api.entity.living.player.Player
+
+Entity AI (not to be confused with actual AI) are fixed goals (defeat enemies) that an entity tries to achieve using
```
Comment (Inscrutable) [inline source/plugin/entities/ai.rst]: "AI" is singular so "are" is weird. Perhaps "is a set of" better expresses the meaning?
Should "defeat enemies" be prefixed by "e.g." ? Same with "go to the enemy ..."

File: source/plugin/entities/ai.rst [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+smaller scaled tasks (go to enemy and hit him with your weapon). The following sections will describe what goals, tasks
```
Comment (Inscrutable) [inline source/plugin/entities/ai.rst]: "what *the* goals, ..."
"the entity AI" seems odd, like it's missing a noun like "system" or "API".
"how you *can* change the ..."
Comma needed after "entities", and "last but not least" should also be contained in commas or parentheses; or we could skip that bit.

File: source/plugin/entities/ai.rst [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+After the cast we can also check whether the AI is actually enabled. Some plugins disable it to simulate statues, guards
+or shops.
+
+Goals
+-----
+
+In vanilla Minecraft there are two :javadoc:`GoalTypes {types}` of :javadoc:`Goal`\s.
+
+The first one is the default one: :javadoc:`GoalTypes#NORMAL {NORMAL}`. This part of the AI controls the behavior of the
+entity. This goal contains tasks such as attacking a target using a melee or ranged weapon and looking idly around the
```
Comment (Inscrutable) [inline source/plugin/entities/ai.rst:52]: comma needed after "weapon"

File: source/plugin/entities/ai.rst [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+area, if there are no enemies nearby. If we want to override or modify the AI we will usually change this goal and it's
```
Comment (Inscrutable) [inline source/plugin/entities/ai.rst]: Optional comma after AI. 
If we're being consistent then that should that really be "its" not "it's" (The Baker's Apostrophe problem).

File: source/plugin/entities/ai.rst [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+
+    Agent entity = ...;
+
+    Optional<Goal<Agent>> normalGoal = entity.getGoal(GoalTypes.NORMAL);
+    if (normalGoal.isPresent()) {
+        normalGoal.get().clear();
+    }
+
+An entity with an empty goal will not move by itself, however it will complete its current action/movement and might
+take an idle pose (look straight ahead). The AI does not affect any sound effects that an entity might play randomly.
```
Comment (Inscrutable) [inline source/plugin/entities/ai.rst]: "i.e. look straight ahead"

File: source/plugin/entities/ai.rst [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+    transfer the tasks from one entity to another.
+
+.. note::
+    
+    Any changes made to an entities AI will be lost after a server restart or after the entity was unloaded.
+
+Tasks
+-----
+
+The behavior of many entity types is similar to each other. For this reason was the entity AI split into multiple
```
Comment (Inscrutable) [inline source/plugin/entities/ai.rst]: "For this reason, the entity API was split ..."

File: source/plugin/entities/ai.rst [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+smaller and reusable parts called :javadoc:`AITask`\s. Each of the ``AITask``\s represents a single behavior trait of an
+entity such as :javadoc:`AITaskTypes#WATCH_CLOSEST {WATCH_CLOSEST}`, which makes the entity look at the nearest matching
+entity, or :javadoc:`AITask#AVOID_ENTITY {AVOID_ENTITY}`, which makes the entity flee from certain matching entities.
+
+.. note::
+
+    Vanilla Minecraft itself provides a huge variety of ``AITask``\s, however most of them are not yet accessible via
+    the API.
+
+Adding additional AITasks
```
Comment (Inscrutable) [inline source/plugin/entities/ai.rst]: Title Case.

File: source/plugin/entities/ai.rst [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+    WatchClosestAITask watchClosestAiTask = WatchClosestAITask.builder()
+            .chance(1)
+            .maxDistance(30)
+            .watch(Player.class)
+            .build(entity);
+    goal.addTask(0, watchClosestAiTask);
+
+In this example we create an :javadoc:`WatchClosestAITask` using the associated builder. Using the build we define that
+the chance of this goal triggering is 100%. Thus whenever a :javadoc:`Player` gets closer to the entity than 30 blocks
+then the entity will look at him. Also we assign zero as a priority, which is a high priority, thus this goal takes
```
Comment (Inscrutable) [inline source/plugin/entities/ai.rst]: him -> them (gender neutral)
Remove Also, and just start the sentence with "We ..."
"takes precedence *above* almost all other tasks"

File: source/plugin/entities/ai.rst [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+precedence about almost all other tasks.
+
+.. note::
+
+    Minecraft uses low values as high priority and high values as low priority. By default Minecraft uses priority
+    values from zero to roughly ten.
+
+Removing certain AITasks
```
Comment (Inscrutable) [inline source/plugin/entities/ai.rst]: Title Case

File: source/plugin/entities/ai.rst [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+The AI API as well as most other parts of the SpongeAPI utilize events. You can read more about events
+:doc:`here </plugin/event/index>`.
+
+The AI API itself makes use of the following 3 events:
+
+* :javadoc:`AITaskEvent.Add`
+* :javadoc:`AITaskEvent.Remove`
+* :javadoc:`SetAITargetEvent`
+
+The ``AITaskEvent.Add`` event is published whenever a new ``AITask`` has been added to a ``Goal``, likewise is the
```
Comment (Inscrutable) [inline source/plugin/entities/ai.rst]: "... likewise, the ``AITaskEvent.Remove`` event *is* published if ..."

File: source/plugin/entities/custom-ai.rst [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+* ``void reset()``
+
+The following image roughly describes the method execution order:
+
+.. image:: /images/ai-execution-order.png
+    :align: center
+    :alt: The method execution order of the ai task
+
+We need the constructor to set any parameters we want our ``AITask`` to have and configure some base options for it. In
+this case its setting the ``AITaskType`` and configuring the mutex bits along with an ``entityFilter``
```
Comment (Inscrutable) [inline source/plugin/entities/custom-ai.rst]: its -> it's
move comma from after `Predicate` to after "bits"

File: source/plugin/entities/custom-ai.rst [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+        super(TYPE);
+        this.entityFilter = Preconditions.checkNotNull(entityFilter);
+        ((EntityAIBase) (Object) this).setMutexBits(MUTEX_FLAG_MOVE);
+    }
+
+Unfortunately we have to resort to dirty casts to plain Minecraft classes. This is also the reason we have to setup our
+workspace with MCP mappings.
+
+After that we continue with the first set of methods: ``canRunConcurrentWith`` and ``canBeInterrupted``. They are very
+easy to implement. For the first one we are going to rely on the default Minecraft logic. However this once again
```
Comment (Inscrutable) [inline source/plugin/entities/custom-ai.rst]: Either put a comma after However, or just replace with "... logic, but this ,,,"

File: source/plugin/entities/custom-ai.rst [hunk 16]
```diff
@@ -0,0 +1,233 @@
...
+        getNavigator().tryMoveToEntityLiving((net.minecraft.entity.Entity) this.optTarget.get(), MOVEMENT_SPEED);
+    }
+
+The owner's :javadoc:`Random` is usually a good source of randomness in those tasks, but any other ``Random`` would work
+as well. The sole reason the :javadoc:`Entity` has a random is to avoid having a ``Random`` in every class that
+interacts with the entity.
+
+The search for the target is pretty straight forward. First we get all entities from the world the entity is currently
+in and then we filter that list using the ``entityFilter`` ``Predicate`` that we set in the constructor. After that we
+remove all entities that are too far away. In our case that are 20 blocks, but you can increase this distance some more.
```
Comment (Inscrutable) [inline source/plugin/entities/custom-ai.rst]: "In our case that are" -> "that is"
"some more" is redundant.

File: source/plugin/entities/custom-ai.rst [hunk 17]
```diff
@@ -0,0 +1,233 @@
...
+Please keep in mind that by increasing the range you also increase the computation time needed for path-finding.
+
+After we have found our target we just have to tell the entity move towards it. The speed is an very important factor
+for this. If the speed is too fast, then the entity might as well accidentally miss the target and end up running back
```
Comment (Inscrutable) [inline source/plugin/entities/custom-ai.rst]: "... the entity might well ..."

File: source/plugin/entities/custom-ai.rst [hunk 18]
```diff
@@ -0,0 +1,233 @@
...
+and forth forever (so make sure that you calculate your acceptance criteria appropriately). If the speed is too slow,
+then the entity will barely move (which looks somewhat strange and makes tasks such as follow X ridiculous).
+
+The above code contains some method calls that have been added to increase the code readability, the following code
```
Comment (Inscrutable) [inline source/plugin/entities/custom-ai.rst]: comma after "readability" should be a semicolon

File: source/plugin/entities/custom-ai.rst [hunk 19]
```diff
@@ -0,0 +1,233 @@
...
+
+    @Override
+    public void reset() {
+        getNavigator().clearPath();
+        this.optTarget = Optional.empty();
+    }
+
+For our ``AITask`` this is very easy. First we check whether the entity lost its path and then whether the entity is
+already close enough. If we return ``false``, then Minecraft will invoke the ``reset()`` method next. There we should
+perform some cleanup. Especially the reference to the entity should be cleaned up, because otherwise we risk a memory
```
Comment (Inscrutable) [inline source/plugin/entities/custom-ai.rst]: "Especially" -> "In particular, "

File: source/plugin/entities/custom-ai.rst [hunk 20]
```diff
@@ -0,0 +1,233 @@
...
+leak. If we return ``true``, then Minecraft will invoke ``update()``, which does nothing in our case because the
+movement controller works independent from the AI and the entity will continue walking towards the target.
```
Comment (Inscrutable) [inline source/plugin/entities/custom-ai.rst]: "independently"

File: source/plugin/entities/custom-ai.rst [hunk 21]
```diff
@@ -0,0 +1,233 @@
...
+
+That's all there is to writing custom ``AITask``\s. Now we can actually use it inside of an ``Entity``, the
+"Adding additional AITasks" chapter on the :doc:`ai` page explains how to do it.
```
Comment (Inscrutable) [inline source/plugin/entities/custom-ai.rst]: Title Case for Titles.
"an ``Entity``, *and* the ..."
final "it." should be "that."

File: source/plugin/entities/custom-ai.rst [hunk 22]
```diff
@@ -0,0 +1,233 @@
...
+
+But why did we call this task ``CreepyCompanionAITask``? There is a simple solution to that. When checking whether the
+``AITask`` should be executed, we don't check whether the entity is already close enough, thus it will move a tiny bit
+towards the target each time. The next step in creepiness would probably be, that the entity only does this if the
```
Comment (Inscrutable) [inline source/plugin/entities/custom-ai.rst]: comma after "be" not needed.

**Review by Inscrutable — COMMENTED** [review]
Pending a more articulate review of the code portions, I have oppressed the wording with more commas and stuff.

File: source/plugin/entities/ai.rst (hunk 4 above)
Comment (liach) [reply source/plugin/entities/ai.rst]: I'd say it's the logic that controls an agent without referring to goals or tasks.

File: source/plugin/entities/ai.rst [hunk 23]
```diff
@@ -0,0 +1,195 @@
...
+    org.spongepowered.api.event.entity.ai.AITaskEvent.Add
+    org.spongepowered.api.event.entity.ai.AITaskEvent.Remove
+    org.spongepowered.api.event.entity.ai.SetAITargetEvent
+    org.spongepowered.api.entity.ai.task.builtin.WatchClosestAITask
+    org.spongepowered.api.entity.living.Agent
+    org.spongepowered.api.entity.living.player.Player
+
+Entity AI (not to be confused with actual AI) are fixed goals (defeat enemies) that an entity tries to achieve using
+smaller scaled tasks (go to enemy and hit him with your weapon). The following sections will describe what goals, tasks
+and other components of the entity AI are, how you change the behavior of entities and last but not least describe how
```
Comment (liach) [inline source/plugin/entities/ai.rst]: `, and, last but not least,`

File: source/plugin/entities/ai.rst (hunk 7 above)
Comment (liach) [reply source/plugin/entities/ai.rst]: Remove the comma before lowercase if. What does `It is tasks` mean?

File: source/plugin/entities/ai.rst [hunk 24]
```diff
@@ -0,0 +1,195 @@
...
+
+    WatchClosestAITask watchClosestAiTask = WatchClosestAITask.builder()
+            .chance(1)
+            .maxDistance(30)
+            .watch(Player.class)
+            .build(entity);
+    goal.addTask(0, watchClosestAiTask);
+
+In this example we create an :javadoc:`WatchClosestAITask` using the associated builder. Using the build we define that
+the chance of this goal triggering is 100%. Thus whenever a :javadoc:`Player` gets closer to the entity than 30 blocks
```
Comment (liach) [inline source/plugin/entities/ai.rst]: ``Thus, the entity will look at the closest :javadoc:`Player` in a range of 30 blocks.``

**Review by gabizou — APPROVED** [review]
Good. No complaints.

**Comment by Zidane** [comment]
Agreed with @gabizou

**Comment by liach** [comment]
Let's gladly accept this.

**Comment by liach** [comment]
Uh, after translating, I found a few minor issues. They are fine

**Pull request merged**